app/townsquare/views.py

```diff
@@ -0,0 +1,216 @@
+
+from dashboard.models import Activity
+from marketing.mails import comment_email
+from ratelimit.decorators import ratelimit
+
+from .models import Comment, Flag, Like, Offer, OfferAction
+from .utils import is_user_townsquare_enabled
+
+
+def get_next_time_available(key):
+    d = timezone.now()
+    if key == 'daily':
+        hours = int(d.strftime('%H'))
+        minutes = int(d.strftime('%M'))
+        d = d + timezone.timedelta(hours=hours) + timezone.timedelta(minutes=minutes)
+    if key == 'weekly':
+        d = d + timezone.timedelta(days=5 - d.weekday())
+    if key == 'monthly':
+        month = int(d.strftime('%m'))
+        year = int(d.strftime('%Y'))
+        year += 1 if month > 11 else 0
+        month += 1
+        d = timezone.datetime(year=year, month=month, day=1)
+    return d
+
+
+def index(request):
+
+    # TODO: temporary until town square is approved for non-staff use
+    if not is_user_townsquare_enabled(request.user):
+        from retail.views import index as regular_homepage
+        return regular_homepage(request)
+
+    # setup tabas
+    tabs = [{
+        'title': "My Tribes",
+        'slug': 'my_tribes',
+    }, {
+        'title': "Everywhere",
+        'slug': 'everywhere',
+    }]
+    for keyword in request.user.profile.keywords:
+        tabs.append({
+            'title': keyword.title(),
+            'slug': f'keyword-{keyword}',
+        })
+    default_tab = 'my_tribes' if request.user.is_authenticated else 'everywhere'
+    tab = request.GET.get('tab', default_tab)
+    if "activity:" in tab:
+        tabs.append({
+            'title': "Search",
+            'slug': tab,
+        })
+
+
+    # get offers
+    offers_by_category = {}
+    for key in ['daily', 'weekly', 'monthly']:
+        next_time_available = get_next_time_available(key)
+        offer = Offer.objects.current().filter(key=key).first()
+        if request.user.is_authenticated:
+            if request.user.profile.offeractions.filter(what='click', offer=offer):
+                offer = None
+        offers_by_category[key] = {
+            'offer': offer,
+            'time': next_time_available.strftime('%Y-%m-%dT%H:%M:%SZ'),
+        }
+
+    is_subscribed = False
+    if request.user.is_authenticated:
+        email_subscriber = request.user.profile.email_subscriptions.first()
+        if email_subscriber:
+            is_subscribed = email_subscriber.should_send_email_type_to('new_bounty_notifications')
+
+    # render page context
+    context = {
+        'title': 'Home',
+        'card_desc': 'View the recent activity on the Gitcoin network',
+        'nav': 'home',
+        'target': f'/activity?what={tab}',
+        'tab': tab,
+        'tabs': tabs,
+        'is_subscribed': is_subscribed,
+        'offers_by_category': offers_by_category,
+    }
+    return TemplateResponse(request, 'townsquare/index.html', context)
+
+
+@ratelimit(key='ip', rate='10/m', method=ratelimit.UNSAFE, block=True)
+@csrf_exempt
```

Comment:
havent figured out how to feed enough csrf tokens to the frontend to make multi comment experience work well. punting for now.

Reply:
So I spent some time looking into this and I believe you can use the same CSRF token multiple times - it only gets invalidated on user login. https://stackoverflow.com/questions/25507514/does-django-csrf-token-must-be-unique-on-every-request

Reply:
fixing this now